src/Bundle.ts

```diff
@@ -104,54 +385,115 @@ export default class Bundle {
-			const entryDirname = dirname(this.entryModule.id);
-			const relativeToEntry = normalize(relative(entryDirname, resolvedId));
+	private setIdentifierRenderResolutions (options: OutputOptions) {
+		const used = blank();
+
+		let dynamicImportMechanism: DynamicImportMechanism;
+		const es = options.format === 'es'
 
-			return isRelative(relativeToEntry)
-				? relativeToEntry
-				: `./${relativeToEntry}`;
+		if (!es) {
+			if (options.format === 'cjs') {
+				dynamicImportMechanism = {
+					left: 'Promise.resolve(require(',
+					right: '))'
+				};
+			} else if (options.format === 'amd') {
+				dynamicImportMechanism = {
+					left: 'new Promise(function (resolve, reject) { require([',
+					right: '], resolve, reject) })'
+				}
+			}
 		}
 
-		return resolvedId;
-	}
+		if (this.graph.dynamicImport) {
+			this.orderedModules.forEach(module => {
+				module.dynamicImportResolutions.forEach((replacement, index) => {
+					const node = module.dynamicImports[index];
 
-	deconflict () {
-		const used = blank();
+					if (!replacement)
+						return;
+
+					if (replacement instanceof Module) {
+						// if we have the module in the bundle, inline as Promise.resolve(namespace)
+						// ensuring that we create a namespace import of it as well
+						if (replacement.bundle === this) {
+							node.setResolution(replacement.namespace(), { left: 'Promise.resolve().then(() => ', right: ')' });
+						// for the module in another chunk, import that other chunk directly
+						} else {
+							node.setResolution(`"${replacement.bundle.id}"`, dynamicImportMechanism);
+						}
+					// external dynamic import resolution
+					} else if (replacement instanceof ExternalModule) {
+						node.setResolution(`"${replacement.id}"`, dynamicImportMechanism);
+					// AST Node -> source replacement
+					} else {
+						node.setResolution(replacement, dynamicImportMechanism);
+					}
+				});
+			});
+		}
 
 		// ensure no conflicts with globals
 		keys(this.graph.scope.variables).forEach(name => (used[name] = 1));
 
 		function getSafeName (name: string): string {
-			while (used[name]) {
-				name += `$${used[name]++}`;
+			let safeName = name;
+			while (used[safeName]) {
+				safeName = `${name}$${used[name]++}`;
 			}
-
-			used[name] = 1;
-			return name;
+			used[safeName] = 1;
+			return safeName;
 		}
 
 		const toDeshadow: Set<string> = new Set();
 
 		this.externalModules.forEach(module => {
-			const safeName = getSafeName(module.name);
-			toDeshadow.add(safeName);
-			module.name = safeName;
-
-			// ensure we don't shadow named external imports, if
-			// we're creating an ES6 bundle
-			forOwn(module.declarations, (declaration, name) => {
-				const safeName = getSafeName(name);
+			if (!es || module.exportsNamespace) {
+				const safeName = getSafeName(module.name);
 				toDeshadow.add(safeName);
-				(<ExternalVariable>declaration).setSafeName(safeName);
+				module.name = safeName;
+			}
+		});
+
+		this.imports.forEach(impt => {
```

Comment:
Maybe this could be extracted as well, just like the `orderedModules` loop below? My hope is that then, this really long function could read more like a script of high level intents where it is easy for people not familiar with the code to quickly identify where they need to look to change certain things.

Reply:
Sometimes inline code is easier to read as high level intents only make sense when you understand the high level picture. Maintaining low-level readability is important too.

Reply:
I agree that there is a line where extraction harms readability more than it helps. Nevertheless
for instance in this case, just extracting two methods, e.g. `deconflictImportedVariables` and `deconflictDeclaredVariables` (or similar) would tell me at a glance where I need to look if one of these two concerns needs to be updated. Instead of heaving to start to read each block at least a little to be sure I am  not mistaken about what the block does. It also prevents others from mixing in other concerns carelessly.
From that extraction, I will also learn that it is probably safe to switch these if I am not happy with the deconflicting order (remember there is a ticket to rework just that).
Furthermore if we extract `getSafeName` (does not need to be into a utility file for now, above the class could be enough), the only dependency of these to the function scope are `used` and `es` so these extracted functions are really rather encapsulated and can easily be reasoned about separately.

> high level intents only make sense when you understand the high level picture

I would actually argue the other way around: Telling the reader high level intents by extracting low level functions actually gives them the chance to understand the high level picture as they are not mentally burdened with the need to understand the details of each step at once.